scripts/gloo2.py

```diff
@@ -0,0 +1,475 @@
+                self._gl.enableVertexAttribArray(attr_handle)
+                self._gl[funcname](attr_handle, *args)
+            else:
+                self._gl.bindBuffer(self._gl.ARRAY_BUFFER, 0)
+                self._gl.disableVertexAttribArray(attr_handle)
+                self._gl[funcname](attr_handle, *args)
+        # Validate. We need to validate after textures units get assigned
+        if not self._validated:
+            self._validated = True
+            self._validate()
+    
+    def _validate(self):
+        # Validate ourselves
+        if len(self._unset_variables):
+            print('Program has unset variables: %s' %
+                        self._unset_variables)
+        # Validate via OpenGL
+        self._gl.validateProgram(self._handle)
+        if not self._gl.getProgramParameter(self._handle, 
+                                            self._gl.VALIDATE_STATUS):
+            print(self._gl.getProgramInfoLog(self._handle))
+            raise RuntimeError('Program validation error')
+    
+    def draw(self, mode, selection):
+        """ Draw program in given mode, with given selection (IndexBuffer or
+        first, count).
+        """
+        if not self._linked:
+            raise RuntimeError('Cannot draw program if code has not been set')
+        # Init
+        check_error(self._gl, 'before draw')
+        # Draw
+        if len(selection) == 3:
+            # Selection based on indices
+            id_, gtype, count = selection
+            if count:
+                self._pre_draw()
+                ibuf = self._gl._objects.get(id_, None)
+                ibuf.activate()
+                self._gl.drawElements(mode, count, gtype, None)
+                ibuf.deactivate()
+        else:
+            # Selection based on start and count
+            first, count = selection
+            if count:
+                self._pre_draw()
+                self._gl.drawArrays(mode, first, count)
+        # Wrap up
+        e = self._gl.getError()
+        check_error(self._gl, 'after draw')
+
+
+class Buffer(GlooObject):
+    _target = None
+    _usage = 35048  # DYNAMIC_DRAW - STATIC_DRAW, STREAM_DRAW or DYNAMIC_DRAW
+    
+    def _create(self):
+        self._handle = self._gl.createBuffer()
+        self._buffer_size = 0
+    
+    def delete(self):
+        self._gl.deleteBuffer(self._handle)
+    
+    def activate(self):
+        self._gl.bindBuffer(self._target, self._handle)
+    
+    def deactivate(self):
+        self._gl.bindBuffer(self._target, 0)
+    
+    # todo: is this not always 32bit? for attributes yes...
+    def set_size(self, nbytes):  # in bytes
+        if nbytes != self._buffer_size:
+            self.activate()
+            self._gl.bufferData(self._target, nbytes, self._usage)
+            self._buffer_size = nbytes
+    
+    #  todo: make possible to also resize here, first init, flag?
+    def set_data(self, offset, data):
```

Comment:
It gets a bit tedious to always have to set both size and data. Options:

* A: Keep it; tedious but clear
* B: set the size automatically the first time that the data is set (init `_buffer_size` with None) 
* C: extra flag to `set_data(self, offset, data, set_size=False)`